processing/src/main/java/io/druid/query/search/search/AutoStrategy.java

```diff
@@ -0,0 +1,56 @@
+package io.druid.query.search.search;
+
+import com.metamx.emitter.EmittingLogger;
+import io.druid.collections.bitmap.ImmutableBitmap;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.segment.QueryableIndex;
+import io.druid.segment.Segment;
+
+import java.util.List;
+
+public class AutoStrategy extends SearchStrategy
+{
+  public static final String NAME = "auto";
+
+  private static final EmittingLogger log = new EmittingLogger(AutoStrategy.class);
+
+  public AutoStrategy(SearchQuery query)
+  {
+    super(query);
+  }
+
+  @Override
+  public List<SearchQueryExecutor> getExecutionPlan(SearchQuery query, Segment segment)
+  {
+    final QueryableIndex index = segment.asQueryableIndex();
+
+    if (index != null) {
+      final ImmutableBitmap timeFilteredBitmap = IndexOnlyStrategy.makeTimeFilteredBitmap(index,
+                                                                                          segment,
+                                                                                          filter,
+                                                                                          interval);
+      final Iterable<DimensionSpec> dimsToSearch = getDimsToSearch(index.getAvailableDimensions(),
+                                                                   query.getDimensions());
+
+      // Choose a search query execution strategy depending on the query.
+      // Index-only strategy is selected when
+      // 1) there is no filter,
+      // 2) the total cardinality is very low, or
+      // 3) the filter has a very low selectivity.
```

Comment:
Thank you for the good suggestion. I agree that cost-based approach is cool. Here are some issues what I'm considering.

* Disk access patterns of cursor-based and index-only strategies look different. The disk access cost can be ignored like you said above only when they are almost same, but I'm not sure that they are. 
* When using cursor-based strategy, some filters can be pushed down into the cursor (preFilters), thereby actually skipping reading non-result data. However, it is difficult to figure out which part of filters are preFilters outside ```QueryableIndexStorageAdapter.makeCursors()```. To do so, some refactoring is needed. This also makes implementing cost-based approach difficult. 

I think we can simply ignore these details at first, and improve later. What do you think?